doc/cascadia/Command-Palette-spec.md

```diff
@@ -0,0 +1,269 @@
+    "action": string,
+    "iconPath": string
+}
+```
+
+Command names should be human-friendly names of actions, though they don't need
+to necessarily be related to the action that it fires. For example, a command
+with `newTab` as the action could have `"Open New Tab"` as the name.
+
+The command will be parsed into a new class, `Command`:
+```c++
+class Command
+{
+    winrt::hstring Name();
+    winrt::TerminalApp::ShortcutAction Action();
+    winrt::hstring IconPath();
+}
+```
+We'll add another structure in GlobalAppSettings to hold all these actions. It
+will just be a `std::vector<Command>` in `GlobalAppSettings`.
+
+We'll need app to be able to turn this vector into a `ListView`, or similar, so
+that we can display this list of actions. Each element in the view will be
+intrinsically associated with the `Command` object it's associated with. In
+order to support this, we'll make `Command` a winrt type that implements
+`Windows.UI.Xaml.Data.INotifyPropertyChanged`. This will let us bind the XAML
+element to the winrt type.
+
+When an element is clicked on in the list of commands, we'll raise the event
+corresponding to that `ShortcutAction`. `AppKeyBindings` already does a great
+job of dispatching `ShortcutActions`, so we'll re-use that. We'll pull the basic
+parts of dispatching `ShortcutAction` callbacks into another class,
+`ShortcutActionDispatch`, with a single `PerformAction(ShortcutAction)` method
+(and events for each action). `AppKeybindings` will be initialized with a
+reference to the `ShortcutActionDispatch` object, so that it can call
+`PerformAction` on it. Aditionally, by having a singular
+`ShortcutActionDispatch` instance, we won't need to re-hook up the
+ShortcutAction keybindings each time we re-load the settings.
+
+In `App`, when someone clicks on an item in the list, we'll get the
+`ShortcutAction` associated with that list element, and call PerformAction on
+the app's `ShortcutActionDispatch`. This will trigger the event handler just the
+same as pressing the keybinding.
+
+## UI/UX Design
+
+We'll add another action that can be used to toggle the visibility of the
+command palette. Pressing that keybinding will bring up the command palette.
+
+**OPEN QUESTION**: Do we want the palette to appear over a single pane, or over
+the entire tab content? In [#Future-considerations], I mention having
+profile-scoped commands. Maybe it makes more sense in that scenario to have the
+palette appear in a particular pane, to help identify that the palette has
+commands for _that_ terminal. It could however clutter the UI quickly, and might
+not be as usable for smaller pane sizes. We could just have the per-profile
+commands added when the palette is opened. The way I have it now is as an
+overlay over the entire tab.
+
+The palette will consist of two main UI elements: a text box for searching for
+commands, and a list of commands.
+
+The list of commands will be populated with all the commands by default. Each

```

Comment:
Question: does it look ok to have icons for only some of our actions? Or are we going to design icons for all of our actions? We can definitely use some from XAML, which'll be nice.

Reply:
I'm thinking we don't _need_ to have icons for every action, like how many menus don't:
![image](https://user-images.githubusercontent.com/18356694/62722138-62e62a00-b9d3-11e9-8296-d15ea30fd3a7.png)

We can always put icons we find appropriate in the defaults, but we'd more than likely need someone to help author some icons for some of our scenarios, if we wanted to have icons for all of the defaults.